base_layer/core/src/proof_of_work/lwma.rs

```diff
@@ -0,0 +1,142 @@
+//  If your coin uses median network time instead of node's time, the "revert to
+//  node time" rule (70 minutes in BCH, ZEC, & BTC) should be reduced to FTL/2
+//  to prevent 33% Sybil attack that can manipulate difficulty via timestamps. See:
+// https://github.com/zcash/zcash/issues/4021
+
+use crate::proof_of_work::difficulty::{Difficulty, DifficultyAjustment};
+use std::{cmp, collections::VecDeque};
+
+const TARGET: u64 = 60; // target time of 1 minute
+const BLOCK_WINDOW: u64 = 150; // we should cap the amount of entries int he VecDeques to this number.
+
+pub struct LWMA {
+    timestamps: VecDeque<u64>,
+    cumulative_difficulties: VecDeque<Difficulty>,
+}
+
+impl LWMA {
+    pub fn new() -> LWMA {
+        LWMA {
+            timestamps: VecDeque::new(),
+            cumulative_difficulties: VecDeque::new(),
+        }
+    }
+}
+
+impl DifficultyAjustment for LWMA {
+    fn add(&mut self, timestamp: u64, cum_difficulty: Difficulty) {
+        self.timestamps.push_back(timestamp);
+        self.cumulative_difficulties.push_back(cum_difficulty);
+        if self.timestamps.len() as u64 > BLOCK_WINDOW + 1 {
+            self.timestamps.remove(0);
+            self.cumulative_difficulties.remove(0);
+        }
+    }
+
+    fn get_difficulty(&self) -> Difficulty {
+        LWMA(&self.timestamps, &self.cumulative_difficulties)
+    }
+}
+
+#[allow(non_snake_case)]
+fn LWMA(timestamps: &VecDeque<u64>, cumulative_difficulties: &VecDeque<Difficulty>) -> Difficulty {
+    if timestamps.len() < 2 {
+        return Difficulty::min();
+    }
+    let n = timestamps.len() as f64;
+
+    // Define a k that will be used to get a proper average after weighting the solvetimes.
+    let k = n * (n + 1.0) * (TARGET as f64) / 2.0;
+    let mut avg_target: u64 = 0;
+    let mut this_timestamp;
+    let mut weighted_times = 0;
+    let mut j: u64 = 0;
+    let mut previous_timestamp = timestamps[0];
+
+    // Loop through N most recent blocks.
+    for i in 1..timestamps.len() {
+        // Prevent solvetimes from being negative in a safe way. It must be done like this.
+        // Do not attempt anything like  if (solvetime < 1) {solvetime=1;}
+        // The +1 ensures new coins do not calculate nextTarget = 0.
+        if timestamps[i] > previous_timestamp {
+            this_timestamp = timestamps[i];
+        } else {
+            this_timestamp = previous_timestamp + 1;
+        }
+
+        // 6*T limit prevents large drops in diff from long solvetimes which would cause oscillations.
+        let solvetime = cmp::min(6 * TARGET, this_timestamp - previous_timestamp);
+
+        // The following is part of "preventing negative solvetimes".
+        previous_timestamp = this_timestamp;
+
+        // Give linearly higher weight to more recent solvetimes.
+        j += 1;
+        weighted_times += solvetime * j;
+
+        avg_target += cumulative_difficulties[i].as_u64();
+    }
+
+    let avg = k / weighted_times as f64;
```

Comment:
K is a f64, rust wont let me devide f64/u64. 
This is done to prevent rounding issues. "avg" becomes the % difference that the difficulty will be adjusted with and this will be between 0-1